text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+
+##### `/cnb/hooks.d/ruby.Dockerfile`
+```
+...
+LABEL io.buildpacks.rebasable=false # default, not needed
+RUN yum install ruby
+...
+```
+
+##### `/cnb/hooks.d/ruby.Dockerfile.buildpacks`
+```
+com.example.ruby
+```
+
+##### `/cnb/hooks.d/node.Dockerfile`
+```
+...
+LABEL io.buildpacks.rebasable=false # default, not needed
+RUN yum install nodejs
+...
+```
+
+##### `/cnb/hooks.d/node.Dockerfile.buildpacks`
+```
+com.example.node
+```
+
+
+### Dockerfiles for Base Images
+
+The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.
+
+Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
+
+### Example Dockerfiles
+
+Dockerfile used to create a runtime base image:
+
+```
+ARG base_image
+FROM ${base_image}
+ARG build_id=0
+
+LABEL io.buildpacks.image.distro=ubuntu
+LABEL io.buildpacks.image.version=18.04
+LABEL io.buildpacks.rebasable=true
+
+ENV CNB_USER_ID=1234
+ENV CNB_GROUP_ID=1235
+
+RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
+  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb
+
+USER ${CNB_USER_ID}:${CNB_GROUP_ID}
+
+COPY genpkgs /cnb/image/genpkgs
+```
+
+`run.Dockerfile` for use with the example `app.run.Dockerfile.out` hook that always installs the latest version of curl:
+```
+ARG base_image
+FROM ${base_image}
+ARG build_id=0
+
+RUN echo ${build_id}
+
+RUN apt-get update && apt-get install -y curl && rm -rf /var/lib/apt/lists/*
+```
+(note: this Dockerfile disables rebasing, as OS package installation is not rebasable)
+
+`run.Dockerfile` for use with the example `app.run.Dockerfile.out` hook that installs a special package to /opt:
+```
+ARG base_image
+FROM ${base_image}
+ARG build_id=0
+
+LABEL io.buildpacks.rebasable=true
+
+RUN curl -L https://example.com/mypkg-install | sh # installs to /opt
+```
+(note: rebasing is explicitly allowed because only a single directory in /opt is created)
+
+### Future: Dynamic Run-time Base Image Selection
+
+The outcomes targeted by https://github.com/buildpacks/rfcs/pull/174 could be achieved with an additional type of hook.
+
+For example, we could introduce a `ref` format that replaces the image instead of extending it:
+```
+/cnb/hooks.d/myhook.run.ref # contains: index.docker.io/example/my-ruby-run-image
```

Comment:
This could simple be achieved by using

```dockerfile
FROM index.docker.io/example/my-ruby-run-image
```
instead?

Reply:
And instead of ref we could use a normal dockerfile hook?

Reply:
I originally disliked this idea because I assumed that `docker build` would fail on unused build args. Given that it does not complain about them, I agree this is sufficient.